src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs

```diff
@@ -154,202 +110,179 @@ public HttpConnection(Uri url, ITransportFactory transportFactory, ILoggerFactor
-                    if (t.IsFaulted || abortException != null)
-                    {
-                        _startTcs.SetException(_abortException ?? t.Exception.InnerException);
-                    }
-                    else if (t.IsCanceled)
-                    {
-                        _startTcs.SetCanceled();
-                    }
-                    else
-                    {
-                        _startTcs.SetResult(null);
-                    }
-                });
+                    Log.SkippingStart(_logger);
+                    return;
+                }
 
-            return _startTcs.Task;
-        }
+                Log.Starting(_logger);
 
-        private async Task<NegotiationResponse> GetNegotiationResponse()
-        {
-            var negotiationResponse = await Negotiate(Url, _httpClient, _logger);
-            _connectionId = negotiationResponse.ConnectionId;
-            _logScope.ConnectionId = _connectionId;
-            return negotiationResponse;
+                await SelectAndStartTransport(transferFormat);
+
+                _started = true;
+                Log.Started(_logger);
+            }
+            finally
+            {
+                _startLock.Release();
+            }
         }
 
-        private async Task StartAsyncInternal(TransferFormat transferFormat)
+        public async Task DisposeAsync() => await DisposeAsyncCore().ForceAsync();
+
+        private async Task DisposeAsyncCore(Exception exception = null)
         {
-            Log.HttpConnectionStarting(_logger);
+            if (_disposed)
+            {
+                return;
+            }
 
+            await _startLock.WaitAsync();
             try
             {
-                var connectUrl = Url;
-                if (_requestedTransportType == TransportType.WebSockets)
+                if (!_disposed && _started)
                 {
-                    // if we're running on Windows 7 this could throw because the OS does not support web sockets
-                    Log.StartingTransport(_logger, _requestedTransportType, connectUrl);
-                    await StartTransport(connectUrl, _requestedTransportType, transferFormat);
-                }
-                else
-                {
-                    var negotiationResponse = await GetNegotiationResponse();
-
-                    // Connection is being disposed while start was in progress
-                    if (_connectionState == ConnectionState.Disposed)
-                    {
-                        Log.HttpConnectionClosed(_logger);
-                        return;
-                    }
+                    Log.DisposingHttpConnection(_logger);
 
-                    // This should only need to happen once
-                    connectUrl = CreateConnectUrl(Url, negotiationResponse.ConnectionId);
+                    // Complete our ends of the pipes.
+                    _transportPipe.Input.Complete(exception);
```

Comment:
This will noop in the HubConnection case right? DisposeAsync should only ever be called once the HubConnection completes reading and writing right? Otherwise you'll see errors about trying to call GetMemory after completing the pipe.

Reply:
Yes, the goal here is to force the pipes to be completed if they weren't already. It's expected that the caller will have already completed the pipes before disposing.